// easescript/lib/stacks/AnnotationDeclaration.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
const BlankScope = require("../scope/BlankScope");
class AnnotationDeclaration extends Stack{
    constructor(compilation,node,scope,parentNode,parentStack){
        scope = new BlankScope( null );
        super(compilation,node,scope,parentNode,parentStack);
        this.isAnnotationDeclaration=true;
        this.body = (node.body || []).map( item=>{
            return Utils.createStack(compilation,item,scope,node,this);
        });
    }
    get name(){
        return this.node.name;
    }
    freeze(){
        super.freeze(this);
        super.freeze(this.body);
        super.freeze(this.scope);
        this.body.forEach(stack=>stack.freeze());
    }
    definition(){
        return null;
    }
    description(){
        return this;
    }
    getArguments(){
        const target=[];
        this.body.map( (item,index)=>{
            if(item.isAssignmentPattern){
                const key = item.left.value();
                const value = item.right.value();
                const assigned = true;
                target.push({key,value,assigned,stack:item});
            }else{
                const key = index;
                const value = item.value();
                const assigned = false;
                target.push({key,value,assigned,stack:item});
            }
        });
        return target;
    }
}

module.exports = AnnotationDeclaration;
