// easescript/lib/stacks/ArrayExpression.js
const Utils = require("../core/Utils");
const Expression = require("./Expression");
const LiteralArrayType = require("../types/LiteralArrayType");
const keySymbol = Symbol("key");
class ArrayExpression extends Expression{
    constructor(compilation,node,scope,parentNode,parentStack){
        super(compilation,node,scope,parentNode,parentStack);
        this.isArrayExpression=true;
        this.elements = node.elements.map( (item)=>{
            return Utils.createStack(compilation,item,scope,node,this);
        });
        this[keySymbol]={};
    }
    freeze(){
        super.freeze(this);
        super.freeze(this.elements);
        this.elements.forEach(stack=>stack.freeze());
    }
    definition(){
        return null;
    }
    attribute(index){
        if( typeof index === 'number'){
            return this.elements[index] || null;
        }
        return null;
    }
    reference(){
        return this;
    }
    referenceItems(){
        return [this];
    }
    description(){
        return this;
    }
    type(){
        return this[keySymbol]._type || (this[keySymbol]._type = new LiteralArrayType( this.getGlobalTypeById("array"), this ) );
    }
    parser(){
        this.elements.forEach( item=>{
            item.parser();
            item.setRefBeUsed();
        });
    }
    checker(){
        this.elements.forEach( item=>{
            item.checker();
        });
    }
}

module.exports = ArrayExpression;
