// easescript/lib/stacks/BlockStatement.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
const BlockScope = require("../scope/BlockScope");
class BlockStatement extends Stack{
    constructor(compilation,node,scope,parentNode,parentStack){
        if( parentStack && 
            !(parentStack.isFunctionDeclaration || 
            parentStack.isFunctionExpression || 
            parentStack.isArrowFunctionExpression || 
            parentStack.isWhenStatement) 
        ){
            scope = new BlockScope(scope);
        }
        super(compilation,node,scope,parentNode,parentStack);
        this.isBlockStatement= true;
        this.body = [];
        for(const item of node.body){
            const stack = Utils.createStack( compilation, item, scope, node, this );
            if( stack ){
                this.body.push( stack );
            }
        };
    }

    freeze(){
        super.freeze(this);
        super.freeze(this.scope);
        super.freeze(this.body);
        this.body.forEach( stack=>stack.freeze() );
    }

    definition(){
        return null;
    }

    parser(){
        if( super.parser() ){
            this.body.forEach( item=>item.parser() );
            return true;
        }
        return false;
    }

    checker(){
        if( super.checker() ){
            this.body.forEach( item=>item.checker() );
            return true;
        }
        return false;
    }
}

module.exports = BlockStatement;
