// easescript/lib/stacks/ContinueStatement.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
class ContinueStatement extends Stack{
    constructor(compilation,node,scope,parentNode,parentStack){
        super(compilation,node,scope,parentNode,parentStack);
        this.isBreakStatement= true;
        this.label = Utils.createStack(compilation,node.label,scope,node,this);
    }

    freeze(){
        super.freeze(this);
        this.label && this.label.freeze();
    }

    getLabelStackByName( name ){
        let labelStack = this.getParentStack( stack=>!!(stack.isLabeledStatement || stack.isFunctionExpression) );
        while( labelStack && labelStack.isLabeledStatement ){
            if( labelStack.label.value() === name ){
                return labelStack;
            }
            labelStack = labelStack.labelParent;
        }
        return null;
    }

    definition(){
        if( this.label ){
            const labelStack = this.getLabelStackByName( this.label.value() );
            if( labelStack ){
                return labelStack.definition();
            }
        }
        return null;
    }

    checker(){
        if( this.label ){
            const labelStack = this.getLabelStackByName( this.label.value() );
            if( !labelStack || !labelStack.isLabeledStatement ){
                this.label.error( 1022 );
            }
        }
        let parent = this.parentStack;
        while( parent ){
            if( parent.isSwitchCase || 
                parent.isDoWhileStatement || 
                parent.isWhileStatement || 
                parent.isForStatement || 
                parent.isForOfStatement || 
                parent.isForInStatement )
            {
               return;
            }else if(parent.isFunctionExpression){
               break; 
            }else{
                parent = parent.parentStack;
            }
        }
        this.error(1022);
    }
}

module.exports = ContinueStatement;
