// easescript/lib/stacks/DoWhileStatement.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
class DoWhileStatement extends Stack{
   constructor(compilation,node,scope,parentNode,parentStack){
      super(compilation,node,scope,parentNode,parentStack);
      this.isDoWhileStatement= true;
      this.condition = Utils.createStack(compilation,node.test,scope,node,this);
      this.body = Utils.createStack(compilation,node.body,scope,node,this);
   }
   freeze(){
      super.freeze();
      this.condition.freeze();
      this.body.freeze();
   }
   definition(){
      return null;
   }
   parser(){
      this.condition.parser();
      this.condition.setRefBeUsed();
      this.body.parser();
   }
   checker(){
      if( !this.condition ){
           this.error(1041);
      }
      const desc = this.condition.description();
      if( desc.isLiteral ){
         const has = this.body.body.some( item=>item.isReturnStatement || item.isBreakStatement);
         if( !has ){
            this.condition.warn(1042)
         }
      }
   }
}

module.exports = DoWhileStatement;
