// easescript/lib/stacks/EnumProperty.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
const EnumType = require("../types/EnumType");
class EnumProperty extends Stack{

    constructor(compilation,node,scope,parentNode,parentStack){
        super(compilation,node,scope,parentNode,parentStack);
        this.isEnumProperty= true;
        this.key = Utils.createStack(compilation,node.key,scope,node,this);
        let init = node.init;
        if(!init){
            init = Object.assign({},node);
            init.type  = "Literal";
            init.value = parentStack.increment++;
            init.raw = `${init.value}`;
            init.start=-1;
            init.end =-1;
            init.loc = null;
        }
        this.init = Utils.createStack(compilation,init,scope,node,this);
    }
    freeze(){
        super.freeze(this);
        super.freeze(this.init);
        super.freeze(this.key);
    }
    definition(){
        const expre = `enum ${this.parentStack.value()}.${this.value()}=${this.init.value()}`;
        return {
            kind:"enum",
            comments:this.comments,
            identifier:this.value(),
            expre:expre,
            location:this.getLocation(),
            file:this.compilation.file,
            context:this
        };
    }
    reference(){
        return this;
    }
    referenceItems(){
        return [this];
    }
    description(){
        return this;
    }
    type(){
        return this._type || (this._type = new EnumType(this.getModuleById("Number"),this,this.parentStack.type()));
    }
    raw(){
        return this.key.raw();
    }
    value(){
        return this.key.value();
    }
}

module.exports = EnumProperty;
