// easescript/lib/stacks/GenericTypeAssignmentDeclaration.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
const GenericType = require("../types/GenericType");
class GenericTypeAssignmentDeclaration extends Stack {
    constructor(compilation,node,scope,parentNode,parentStack){ 
        super(compilation,node,scope,parentNode,parentStack);
        this.isGenericTypeAssignmentDeclaration= true;
        this.left = Utils.createStack(compilation,node.left, scope, node, this);
        this.right = Utils.createStack(compilation,node.right, scope, node, this); 
    }
    freeze(){
        super.freeze(this);
        super.freeze(this.left);
        super.freeze(this.right);
    }
    definition(){
       const type = this.type();
       if( type.isModule ){
           return type.definition();
       }
       const identifier = this.left.value();
       return {
           kind:this.kind,
           comments:this.comments,
           identifier:identifier,
           expre:`(Type) ${type.toString()}`,
           location:this.left.getLocation(),
           file:this.compilation.file,
           context:this
       };
    }
    error(code,...args){
        this.left.error(code,...args);
    }
    warn(code,...args){
        this.left.warn(code,...args);
    }
    description(){
        return this;
    }
    referenceItems(){
        return [this];
    }

    type(){
        if( !this._type ){
            const isFunGeneric = !!(this.parentStack && this.parentStack.parentStack && this.parentStack.parentStack.isFunctionExpression);
            this._type = new GenericType( this, null, this.right.type(), isFunGeneric );
        }
        return this._type;
    }

    checker(){
        this.right.checker();
    }

    parser(){
        this.right.parser();
    }

    value(){
        return this.left.value();
    }

    raw(){
        return this.raw();
    }
}

module.exports = GenericTypeAssignmentDeclaration;
