// easescript/lib/stacks/IfStatement.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
class IfStatement extends Stack{
    constructor(compilation,node,scope,parentNode,parentStack){
        super(compilation,node,scope,parentNode,parentStack);
        this.isIfStatement = true;
        this.condition = Utils.createStack(compilation,node.test,scope,node,this);
        this.consequent = Utils.createStack(compilation,node.consequent,scope,node,this);
        this.alternate  = Utils.createStack(compilation,node.alternate,scope,node,this);
    }
    freeze(){
        super.freeze(this);
        super.freeze(this.condition);
        super.freeze(this.consequent);
        super.freeze(this.alternate);
    }
    definition(){
        return null;
    }
    checker(){
        if( !this.condition ){
            this.error(1041);
        }
        this.consequent && this.consequent.checker();
        if( this.alternate ){
            this.alternate.checker();
        }
    }
    parser(){
        if( this.condition ){
            this.condition.parser();
            this.condition.setRefBeUsed();
        }
        this.consequent && this.consequent.parser();
        if( this.alternate ){
            this.alternate.parser();
        }
    }
}

module.exports = IfStatement;
