// easescript/lib/stacks/LabeledStatement.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
class LabeledStatement extends Stack{

    constructor(compilation,node,scope,parentNode,parentStack){
        super(compilation,node,scope,parentNode,parentStack);
        this.isLabeledStatement= true;
        this.labelChildren = [];
        this.labelParent = null;
        this.label = Utils.createStack( compilation, node.label, scope, node, this );
        this.body =  Utils.createStack( compilation, node.body, scope, node, this );
        const labelStack =parentStack && parentStack.getParentStack( stack=>!!(stack.isLabeledStatement || stack.isFunctionExpression) );
        if( labelStack && labelStack.isLabeledStatement ){
            this.labelParent = labelStack;
            labelStack.labelChildren.push(this);
        }
    }

    freeze(){
        super.freeze(this);
        super.freeze(this.label);
        super.freeze(this.labelChildren);
        this.body.freeze();
    }

    definition(){
        return {
            kind:"label",
            comments:this.label.comments,
            identifier:this.label.value(),
            expre:null,
            location:this.label.getLocation(),
            file:this.compilation.file,
            context:this
        };
    }

    parser(){
        this.body.parser();
    }

    checker(){
        this.body.checker();
    }
}

module.exports = LabeledStatement;
