// easescript/lib/stacks/LogicalExpression.js
const Utils = require("../core/Utils");
const Expression = require("./Expression");
class LogicalExpression extends Expression{
     constructor(compilation,node,scope,parentNode,parentStack){
          super(compilation,node,scope,parentNode,parentStack);
          this.isLogicalExpression= true;
          this.left = Utils.createStack( compilation, node.left, scope, node,this );
          this.right = Utils.createStack( compilation, node.right, scope, node,this );
     }
     freeze(){
          super.freeze();
          this.left.freeze();
          this.right.freeze();
     }
     definition(){
          return null;
     }
     description(){
          return this;
     }
     type(){
          const operator = this.node.operator;
          const code = operator.charCodeAt(0);
          if( code === 38 ){
               return this.right.type();
          }
          return this.left.type();
     }
     parser(){
          this.left.parser();
          this.left.setRefBeUsed();
          this.right.parser();
          this.right.setRefBeUsed();
     }
     checker(){
          this.left.checker();
          this.right.checker();
     }
}
module.exports = LogicalExpression;
