// easescript/lib/stacks/SequenceExpression.js
const Utils = require("../core/Utils");
const Expression = require("./Expression");
class SequenceExpression extends Expression{
     constructor(compilation,node,scope,parentNode,parentStack){
          super(compilation,node,scope,parentNode,parentStack);
          this.isSequenceExpression= true;
          this.expressions = node.expressions.map( item=>Utils.createStack( compilation, item, scope, node, this ) );
     }
     freeze(){
          super.freeze();
          super.freeze( this.expressions );
          (this.expressions || []).forEach( stack=>stack.freeze() );
      }
     definition(){
          return null;
     }
     description(){
          return this.expressions.length > 0 ? this.expressions[ this.expressions.length-1 ] : this;
     }
     reference(){
          const description = this.description();
          return description !== this ? description.reference() : this;
     }
     referenceItems(){
          const description = this.description();
          return description !== this ? description.referenceItems() : [this];
     }
     type(){
          return this.expressions.length > 0 && this.expressions[ this.expressions.length-1 ].type();
     }
     parser(){
          this.expressions.forEach( item=>{
               item.parser();
               item.setRefBeUsed();
          });
     }
     checker(){
          this.expressions.forEach( item=>item.checker() );
     }
}

module.exports = SequenceExpression;
