// easescript/lib/stacks/SpreadElement.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
class SpreadElement  extends Stack{
    constructor(compilation,node,scope,parentNode,parentStack){
        super(compilation,node,scope,parentNode,parentStack);
        this.isSpreadElement=true;
        this.argument = Utils.createStack( compilation, node.argument, scope, node,this );
    }
    freeze(){
        super.freeze();
        this.argument.freeze();
    }
    definition(){
        return null;
    }
    description(){
        return this.argument.description();
    }
    reference(){
        return this.argument.reference();
    }
    referenceItems(){
        return this.argument.referenceItems();
    }
    key(){
        return this.argument.key();
    }
    type(){
        return this.argument.type();
    }
    checker(){
        this.argument.checker();
        const type = this.type();
        if( this.parentStack.isArrayExpression ){
            const arrayType = this.getGlobalTypeById("array");
            const iteratorType = this.getGlobalTypeById("Iterator");
            if( !( arrayType.is( type ) || type.is(iteratorType) ) ){
                this.error(1073,this.argument.value())
            }  
        }else if( this.parentStack.isObjectExpression ){
            const objectType = this.getGlobalTypeById("object");
            if( !objectType.is(type) ){
                this.error(1074,this.argument.value())
            }
        }
    }
    parser(){
        this.argument.parser();
        this.argument.setRefBeUsed();
    }
    value(){
        return this.argument.value();
    }
    raw(){
        return this.argument.raw();
    }
    error(code,...args){
        this.argument.error(code,...args);
    }
    warn( message ){
        this.argument.error(code,...args);
    }
}

module.exports = SpreadElement;
