// easescript/lib/stacks/SwitchStatement.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
const BlockScope = require("../scope/BlockScope");
class SwitchStatement  extends Stack {
    constructor(compilation,node,scope,parentNode,parentStack){ 
        super(compilation,node,scope,parentNode,parentStack);
        this.isSwitchStatement=true;
        this.condition = Utils.createStack(compilation, node.discriminant, scope, node,this );
        scope = new BlockScope(scope);
        this.cases = node.cases.map( item=>{
           return Utils.createStack( compilation, item, scope, node, this );
        });
    }
    freeze(){
        super.freeze();
        super.freeze( this.cases );
        super.freeze( this.scope );
        this.condition.freeze();
        this.cases.forEach( stack=>stack.freeze() );
    }
    definition(){
        return null;
    }
    parser(){
        this.condition.parser();
        this.condition.setRefBeUsed();
        this.cases.forEach( item=>item.parser() );
    }
    checker(){
        this.condition.checker();
        this.cases.forEach( item=>item.checker() );
    }
}

module.exports = SwitchStatement;
