// easescript/lib/stacks/TypeUnionDefinition.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
const UnionType = require("../types/UnionType");
const keySymbol = Symbol("key");
class TypeUnionDefinition extends Stack {
    constructor(compilation,node,scope,parentNode,parentStack){ 
        super(compilation,node,scope,parentNode,parentStack);
        this.isTypeUnionDefinition= true;
        this.elements = node.elements.map( item=>{
            const stack = Utils.createStack(compilation,item,scope,node,this);
            return stack;
        });
        this[keySymbol]={};
    }
    freeze(){
        super.freeze();
        super.freeze( this.elements );
        this.elements.forEach( stack=>stack.freeze() );
    }
    definition(){
      return null;
    }
    description(){
        return this;
    }
    referenceItems(){
        return [this];
    }
    type(){
        if( !this[keySymbol]._type ){
            this[keySymbol]._type = new UnionType( this.elements , this);
        }
        return this[keySymbol]._type;
    }
    parser(){
        this.elements.forEach( item=>item.parser() );
    }
    checker(){
        this.elements.forEach( item=>item.checker() );
    }
    value(){
        const elements = this.elements.map( item=>{
            return item.value()
        });
        return `${elements.join("|")}`;
    }
    raw(){
        return this.value();
    }
}

module.exports = TypeUnionDefinition;
