// easescript/lib/stacks/TypeTupleDefinition.js
const Stack = require("../core/Stack");
const Utils = require("../core/Utils");
const TupleType = require("../types/TupleType");
const keySymbol = Symbol("key");
class TypeTupleDefinition extends Stack {
    constructor(compilation,node,scope,parentNode,parentStack){ 
        super(compilation,node,scope,parentNode,parentStack);
        this.isTypeTupleDefinition= true;
        this.prefix = Utils.createStack(compilation,node.prefix,scope,node,this);
        this.elements = (node.elements||[]).map( item=>{
            const stack = Utils.createStack(compilation,item,scope,node,this);
            return stack;
        });
        this[keySymbol]={};
    }
    freeze(){
        super.freeze();
        super.freeze( this.elements );
        this.prefix && this.prefix.freeze();
        this.elements.forEach( stack=>stack.freeze() );
    }
    definition(){
       const type = this.type();
       return {
           kind:this.kind,
           comments:this.comments,
           expre:`(type) ${type.toString()}`,
           location:(this.prefix||this).getLocation(),
           file:this.compilation.file,
           context:this
       };
    }
    description(){
        return this;
    }
    reference(){
        return this;
    }
    referenceItems(){
        return [this];
    }
    type(){
        if( !this[keySymbol]._type ){
            const elem =  this.prefix ? [this.prefix] : this.elements;
            if( elem.length > 0 ){
                this[keySymbol]._type = new TupleType(
                    this.getGlobalTypeById("Array"),
                    elem,
                    this
                );
            }else{
                this[keySymbol]._type = this.getGlobalTypeById("array");
            }
        }
        return this[keySymbol]._type;
    }

    checker(){
        this.prefix && this.prefix.checker();
        this.elements.forEach( item=>item.checker() );
    }

    parser(){
        if( this.prefix  && this.elements.length > 0){
            this.error(1086, this.prefix.raw() )
        }
        this.prefix && this.prefix.parser();
        this.elements.forEach( item=>item.parser() );
        const restElement = this.elements.find( item=>item.restElement);
        if( restElement && restElement !== this.elements[ this.elements.length-1 ] ){
            this.error(1077);
        }
    }
}

module.exports = TypeTupleDefinition;
