// easescript/lib/stacks/TypeAssertExpression.js
const Utils = require("../core/Utils");
const Expression = require("./Expression");
class TypeAssertExpression extends Expression{

     constructor(compilation,node,scope,parentNode,parentStack){
          super(compilation,node,scope,parentNode,parentStack);
          this.isTypeAssertExpression= true;
          this.left = Utils.createStack( compilation, node.left, scope, node,this );
          this.right = Utils.createStack( compilation, node.right, scope, node,this );
     }
     freeze(){
          super.freeze();
          this.left.freeze();
          this.right.freeze();
     }
     definition(){
          return null;
     }
     description(){
          return this;
     }
     reference(){
          return this.left.reference();
     }
     referenceItems(){
          return this.left.referenceItems();
     }
     type(){
          return this.right.type();
     }
     parser(){
          this.left.parser();
          this.right.parser();
     }
     checker(){
          this.left.checker();
          this.right.checker();
     }
}

module.exports = TypeAssertExpression;
