// easescript/lib/stacks/UnaryExpression.js
const Utils = require("../core/Utils");
const Expression = require("./Expression");
class UnaryExpression extends Expression {
    constructor(compilation,node,scope,parentNode,parentStack){ 
        super(compilation,node,scope,parentNode,parentStack);
        this.isUnaryExpression= true;
        this.argument = Utils.createStack( compilation, node.argument, scope, node,this );
    }
    freeze(){
        super.freeze();
        this.argument.freeze();
    }
    definition(){
        return null;
    }
    description(){
        return this;
    }
    parser(){
        this.argument.parser();
        this.argument.setRefBeUsed();
    }
    checker(){
        const operator = this.node.operator;
        const code = operator.charCodeAt(0);
        if( code === 45 || code === 43 ){
            const numberType = this.getGlobalTypeById("Number");
            const type = this.argument.type();
            if( type && !type.isAnyType && !numberType.is( type ) ){
                this.error(1087, this.argument.raw() );
             }
            return type;
        }
        return this.argument.checker();
    }

    value(){
        return super.raw();
    }

    type(){
        const operator = this.node.operator;
        const code = operator.charCodeAt(0);
        if( code === 33 ){
            return this.getGlobalTypeById("boolean");
        }
        if( code === 45 || code === 43 ){
            const numberType = this.getGlobalTypeById("Number");
            const type = this.argument.type();
            if( type.isAnyType || !numberType.is( type ) ){
                return this.getGlobalTypeById("int");
            }
            return type;
        }
        if( operator === 'typeof' ){
            return this.getGlobalTypeById("string");
        }
        return this.argument.type();
    }
}

module.exports = UnaryExpression;
