// easescript/lib/stacks/VariableDeclaration.js
const Utils = require("../core/Utils");
const Stack = require("../core/Stack");
class VariableDeclaration extends Stack {
    constructor(compilation,node,scope,parentNode,parentStack){ 
        super(compilation,node,scope,parentNode,parentStack);
        this.isVariableDeclaration= true;
        this.kind = node.kind;
        this.declarations = node.declarations.map( item=>{
            return Utils.createStack(compilation,item, scope, node,this);
        });
        this.flag = false;
        switch( parentNode && parentNode.type ){
            case "ForStatement":
            case "ForInStatement":
            case "ForOfStatement":
                this.flag=true;
            break;    
        }
    }
    freeze(){
        super.freeze();
        super.freeze( this.declarations );
        this.declarations.forEach( stack=>stack.freeze() );
    }
    definition(){
        return null;
    }
    reference(){
        return this.declarations[0].reference();
    }
    referenceItems(){
        return this.declarations[0].referenceItems();
    }
    value(){
        return this.declarations[0].value();
    }
    raw(){
        return this.declarations[0].raw();
    }
    warn(code,...args){
        this.declarations[0].warn(code,...args);
    }
    error(code,...args){
        this.declarations[0].error(code,...args);
    }
    parser(){ 
        this.declarations.forEach( item=>item.parser() );
    }
    checker(){ 
        this.declarations.forEach( item=>item.checker() );
    }
}

module.exports = VariableDeclaration;
