// easescript/lib/types/EnumType.js
const Type = require("./Type");
class EnumType extends Type{
    constructor(inherit,target,owner=null){
        super("$EnumType",inherit);
        this.target = target;
        this.isEnumType = true;
        this.owner = owner;
        this.acceptProperties = [];
    }
    accept( property ){
        this.acceptProperties.push( property );
    }
    attribute( property ){
        return this.target.attribute(property);
    }
    is( type ){
        if( !type || !(type instanceof Type) )return false;
        if( type.isGenericValueType ){
            type = type.value;
            if( !type ){
                return true;
            }
        }
        if( type === this || type.isNullableType || type.isAnyType )return true;
        if( this.owner ){
            return this.inherit.is( type );
        }
        return type.isEnumType && type.owner === this;
    }
    toString(){
       if(this.owner){
          if( this.owner.isEnumType ){
              return `${this.owner.target.key.value()}.${this.target.value()}`;
          }else{
              return `${this.owner.id}.${this.target.value()}`;
          }
       }
       let properties = [];
       const inherit = this.inherit;
       if( inherit && inherit.isModule && inherit.isEnum ){
            properties = Object.keys(inherit.methods);
       }else{
            properties = this.target.properties.map( item=>item.value() );
       }
       return `enum {${properties.join(",")}}`;
    }
}
module.exports = EnumType;
