// easescript/lib/core/Scope.js
const EventDispatcher = require("./EventDispatcher.js");
module.exports = class Scope extends EventDispatcher{
    constructor( parent ){
        super();
        this.parent = parent;
        this.children=[];
        this.level = 0;
        if( parent ){
            parent.children.push( this );
            this.level = parent.level+1;
            this.asyncParentScopeOf = parent.async ? parent : parent.asyncParentScopeOf;
        }
        if( parent && parent.topDeclarations ){
            this.topDeclarations = parent.topDeclarations;
        }else{
            this.topDeclarations = new Set();
        }
        this.declarations=new Map();
    }
    
    removeChild( childScope ){
        const index = this.children.indexOf( childScope );
        if( index >= 0 ){
            return this.children.splice(index,1);
        }
        return null;
    }

    hasChildDeclared(name){
        const check = item=>{
            if( item.isDefine( name ) ){
                return true;
            }else{
                return item.children.some( check );
            }
        }
        return check( this );
    };

    generateVarName( name, flag=false){
        let count = 0;
        let ref = name;
        while( (flag === false && this.topDeclarations.has(ref)) || (flag===true && this.hasChildDeclared(ref)) ){
            count++;
            ref=name+count;
        }
        this.define(ref,this);
        return ref;
    }
    
    getKeys(){
       return []
    }

    type( name ){
        return false;
    }

    getScopeByType( name ){
        let obj = this;
        while( obj && obj instanceof Scope && !obj.type( name ) ){
            obj = obj.parent;
        }
        return obj;
    }

    define(name, stack){
        if( stack === void 0 ){
            let has = this.declarations.has( name );
            if( !has ){
                let parentScope = this.parent;
                while( parentScope && parentScope instanceof Scope ){
                    if( parentScope.declarations.has(name) ){
                        return parentScope.declarations.get(name);
                    }else{
                        parentScope = parentScope.parent;
                    }
                }
                return null;
            }
            return this.declarations.get(name);
        }

        this.declarations.set(name,stack);
        if( this.topDeclarations ){
            this.topDeclarations.add(name);
        }
    }

    getDefine(name, context){
        let has = this.declarations.has( name );
        if( !has ){
            let parentScope = this.parent;
            while( parentScope && parentScope instanceof Scope ){
                if( parentScope.declarations.has(name) ){
                    return parentScope.declarations.get(name);
                }else{
                    if( context && (parentScope.type(context) || (context==="block" && parentScope.type("function") ) )){
                        return false;
                    }
                    parentScope = parentScope.parent;
                }
            }
            return null;
        }
        return this.declarations.get(name);
    }

    getKeys(context){
        let keys = Array.from(this.declarations.keys());
        let parent = this.parent;
        while( parent ){
            const is =!context || context.some( name=>!!parent.type(name) );
            if( is ){
                keys = keys.concat( Array.from(parent.declarations.keys()) );
                parent = parent.parent;
            }
            break;
        }
        return keys;
    }

    isDefine( name, context ){
        let parentScope = this;
        while( parentScope && parentScope instanceof Scope){
            if( parentScope.declarations.has(name) ){
                return true
            }else{
                if( context && (parentScope.type(context) || (context==="block" && parentScope.type("function") ) )){
                    return false;
                }
                parentScope = parentScope.parent; 
            }
        }
        return false;
    }
};
